src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php

```diff
@@ -79,14 +82,38 @@ public function decode(array $encodedEnvelope): Envelope
         }
 
         try {
-            $message = $this->serializer->deserialize($encodedEnvelope['body'], $encodedEnvelope['headers']['type'], $this->format, $context);
+            $message = $this->deserialize($encodedEnvelope, $context);
         } catch (ExceptionInterface $e) {
             throw new MessageDecodingFailedException('Could not decode message: '.$e->getMessage(), $e->getCode(), $e);
         }
 
         return new Envelope($message, $stamps);
     }
 
+    private function deserialize(array $encodedEnvelope, array $context)
+    {
+        $resolver = $context[self::TYPE_RESOLVER] ?? new HeaderTypeResolver();
+
+        if ($resolver instanceof TypeResolverInterface) {
+            $type = $resolver->resolve($encodedEnvelope);
+
+            return $this->serializer->deserialize($encodedEnvelope['body'], $type, $this->format, $context);
+        }
+
+        if ($resolver instanceof DecodedAwareTypeResolverInterface) {
+            if (!$this->serializer instanceof DecoderInterface || !$this->serializer instanceof DenormalizerInterface) {
```

Comment:
This is a BC break. 

If I currently pass a `Symfony\Component\Serializer\SerializerInterface` as first argument to the `Symfony\Component\Messenger\Transport\Serialization\Serializer`, then it will stop working if we merge this PR.

Reply:
Hm.. I might be wrong. Since the default resolver is a `TypeResolverInterface` this code will never be reached.

Reply:
Yes indeed, this occurs only with `DecodedAwareTypeResolverInterface` that is not the default resolver.